### DISCUSSION TITLE
Building dynamic library libsodium.so for iOS

### DISCUSSION THREAD

**Discussion by FeodorFitsner** [jedisct1/libsodium#1436] [Q&A]
When building libsodium for iOS with `./configure` and `make` I'm getting only static `libsodium.a` in "lib" folder, but I need a dynamic library `libsodium.so` as well. Is there a way to build that for iOS? I've tried `--disable-static` but it didn't work.

**Comment by matejvasek** [comment]
Maybe related maybe not:
When I was trying to cross-compile on my desktop (Linux amd64) for my e-book reader (Linux arm7) I was getting only the static library `libsodium.a` too. When I properly set up `qemu` emulation by setting `QEMU_LD_PREFIX` to the cross-compiler sysroot it produced also the `libsodium.so` file. As if the `make` also executed some binaries produced for arm7 in the process which was obviously problem on amd64.

This can be also seen when running `make check`.
```
FAIL: aead_aegis128l
====================

qemu-arm-static: Could not open '/lib/ld-linux.so.3': No such file or directory
FAIL aead_aegis128l (exit status: 255)
```
EDIT: hmm trying it today I cannot reproduce missing *.so. The *.so file is produced even without emulation set up. However the tests `make test` do require emulation.

EDIT 2: What about: https://libsodium.gitbook.io/doc/installation#cross-compiling-to-apple-devices ?

**Reply by jedisct1** [reply]
`zig build -Dtarget=arm-linux -Doptimize=ReleaseFast` is the easiest way to cross compile to arm-linux.
Individual tests are put in the `zig-out/bin` directory, that you can run with emulation later if you wish, with `./run.sh`.

**Comment by matejvasek** [comment]
Also have you tried `--enable-shared` ?

**Comment by jedisct1** [comment]
Libraries for Apple devices can be created with the `./dist-build/apple-xcframework.sh` script. It doesn't do anything special.
The xcframework only includes static libraries, but `.dylib` files can be found in the `libsodium-apple`  subdirectories.

**Comment by FeodorFitsner** [comment]
Thanks @jedisct1, studying `apple-xcframework.sh` helped. I was using `--host=arm64-apple-ios` triplet and only `libsodium.a` was built while using `--host=arm-apple-darwin23` yields both `libsodium.a` and `libsodium.dylib`.
